### PR TITLE
feat: migrate harvester action API to subresource

### PR DESCRIPTION
**Pull request by Yu-Jack**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
Currently, there are fews APIs don't follow kubernetes rules. In order to make use of kubernetes API format, we should migrate those APIs.

**Solution:**
HEP https://github.com/harvester/harvester/pull/5885

We could use Aggregated API Server (API Service) to solve this issue. Need to define a `subresources.harvesterhci.io` to proxy request to our harvester server. It's similar with [kubevirt subresource](https://kubevirt.io/api-reference/v1.2.0/operations.html#_v1start).

Create another common interface to deal with two different sources, this is a simple visual flow of objects.

image

**Related Issue:**
https://github.com/harvester/harvester/issues/1078

**Test plan:**

Before testing, should check there is a harvester apiservice:

```yaml
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  name: v1beta1.subresources.harvesterhci.io
spec:
  group: subresources.harvesterhci.io
  groupPriorityMinimum: 1000
  service:
    name: harvester
    namespace: harvester-system
    port: 8443
  version: v1beta1
  versionPriority: 15
  insecureSkipTLSVerify: true
```

Before testing, these are example action API lists we currently use:
- GET `/v1/harvester/harvesterhci.io.virtualmachineimages/default/image-g88lp/download`
- POST `/v1/harvester/kubevirt.io.virtualmachines/default/test?action=stop`
- POST `/v1/harvester/nodes/jacknode?action=cordon`
- GET `/v1/harvester/upgradeslogs/default/test/download`
- POST `/v1/harvester/persistentvolumeclaims/default/test-disk-0-rlnlk?action=snapshot`

We should test those APIs by using subresource APIs, for example:

- GET `/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/virtualmachineimages/ubuntu-focal/download`
- POST `/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/virtualmachines/test/stop`
- POST `/apis/subresources.harvesterhci.io/v1beta1/nodes/jacknode/uncordon`
- GET `/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/upgradelogs/test/download`
- POST `/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/persistentvolumeclaims/test/snapshot`
- `/apis/subresources.harvesterhci.io/v1beta1/namespaces/default/snapshot.storage.k8s.io.volumesnapshots/test/restore`

Then you could also use following REST config to invoke subresource:

```go
harvesterCopyConfig := rest.CopyConfig(server.RESTConfig)
harvesterCopyConfig.GroupVersion = &k8sschema.GroupVersion{Group: "subresources.harvesterhci.io", Version: "v1beta1"}
harvesterCopyConfig.APIPath = "/apis"
harvesterCopyConfig.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
harvesterSubresourceClient, err := rest.RESTClientFor(harvesterCopyConfig)

h.harvesterSubresourceClient.Post().Resource("nodes").SubResource("uncordon").Name(node.Name).Do(context.Background())
h.harvesterSubresourceClient.Post().Namespace("default").Resource("virtualmachines").SubResource("stop").Name("test").Do(context.Background())
```

When testing with wrong resource/name/subresource, will get different status code and message like:

- 422 - `InvalidAction 422: Unsupported virtualmachines/stopa`
- 422 - `InvalidAction 422: Unsupported virtualmachinessss/stop`
- 404 - `NotFound 404: Virtual machine default/testa not found`

Currently, I use rancher/apiserver to generate apierror like we did in action part. But, I think it's still discussable which ways we'd like to respond.

**Actions should migrate**

- [x] image
- [x] node
- [x] upgradelog
- [x] vm
- [x] volume
- [x] volumesnapshot

### CODE REVIEW

File: pkg/api/image/formatter.go [hunk 1]
```diff
@@ -68,31 +97,28 @@ func (h Handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
...
-}
 
-func (h Handler) doPost(action string, rw http.ResponseWriter, req *http.Request) error {
-	switch action {
-	case actionUpload:
-		return h.uploadImage(rw, req)
-	default:
-		return apierror.NewAPIError(validation.InvalidAction, fmt.Sprintf("Unsupported POST action %s", action))
+	if !h.IsMatchedResource(resource, req.Method) {
+		return apierror.NewAPIError(validation.InvalidAction, fmt.Sprintf("Unsupported %s action %s", req.Method, resource.SubResource))
```
Comment (Yu-Jack) [inline pkg/api/image/formatter.go]: We should keep original action and method error message for old APIs.

File: pkg/api/image/formatter.go [hunk 2]
```diff
@@ -53,6 +55,33 @@ type Handler struct {
...
+}
+
+func (h Handler) SubResourceHandler(rw http.ResponseWriter, req *http.Request, resource subresource.Resource) error {
+	switch resource.SubResource {
+	case actionDownload:
+		return h.downloadImage(rw, req)
+	case actionUpload:
+		return h.uploadImage(rw, req)
+	default:
+		return apierror.NewAPIError(validation.InvalidAction, fmt.Sprintf("Unsupported subresource %s", resource.SubResource))
```
Comment (Yu-Jack) [inline pkg/api/image/formatter.go:83]: And this error message is for new subresources APIs.

File: pkg/server/subresource/subresource.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+var (
+	apiService = schema.GroupVersion{Group: "subresources.harvesterhci.io", Version: "v1beta1"}
+
+	VirtualMachineImages   = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "virtualmachineimages"}
+	UpgradeLogs            = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "upgradelogs"}
+	Node                   = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "nodes"}
+	VirtualMachines        = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "virtualmachines"}
+	PersistentVolumeClaims = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "persistentvolumeclaims"}
+	VolumeSnapshots        = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "snapshot.storage.k8s.io.volumesnapshot"}
```
Comment (Yu-Jack) [inline pkg/server/subresource/subresource.go:23]: I also collected all resource which contains action here, it would be more clear.

Comment (Yu-Jack) [inline pkg/server/subresource/subresource.go:23]: I collected all resource which contains action or link here, it'd be more clear. Then, we should consider what resources name we should use.

File: pkg/server/subresource/subresource.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			Namespace:   vars["namespace"],
+		}
+	)
+
+	if vars["namespace"] != "" {
+		resource.Namespaced = true
+		resource.Namespace = vars["namespace"]
+	}
+
+	logrus.Debug("Resource: ", resource)
```
Comment (m-ildefons) [inline pkg/server/subresource/subresource.go:107]: I think it would be a good idea to log any subresource requests. This would help tremendously when dealing with support cases as the service logs would then contain information on what actions happened and when.

File: pkg/api/image/formatter.go [hunk 5]
```diff
@@ -53,6 +54,33 @@ type Handler struct {
...
+func (h Handler) IsMatchedResource(resource subresource.Resource, method string) bool {
+	if resource.Name != subresource.VirtualMachineImages.Resource {
+		return false
+	}
+
+	if resource.SubResource == actionDownload && method == http.MethodGet {
+		return true
+	}
+
+	if resource.SubResource == actionUpload && method == http.MethodPost {
```
Comment (m-ildefons) [inline pkg/api/image/formatter.go]: Shouldn't this be HTTP method `PUT`, since the action is idempotent?

File: pkg/api/node/formatter.go [hunk 6]
```diff
@@ -102,15 +113,31 @@ func (h ActionHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	rw.WriteHeader(http.StatusNoContent)
 }
 
-func (h ActionHandler) do(rw http.ResponseWriter, req *http.Request) error {
-	vars := util.EncodeVars(mux.Vars(req))
-	action := vars["action"]
-	name := vars["name"]
+func (h ActionHandler) IsMatchedResource(resource subresource.Resource, method string) bool {
+	if resource.Name != subresource.Node.Resource || method != http.MethodPost {
```
Comment (m-ildefons) [inline pkg/api/node/formatter.go]: Again, shouldn't this be a HTTP method `PUT` for idempotent actions (such as cordon or uncordon)?
Additionally, list-actions like `listUnhealthyVM` should be a `GET`, right?

File: pkg/api/upgradelog/handler.go [hunk 7]
```diff
@@ -82,6 +83,33 @@ type Handler struct {
...
+		return false
+	}
+
+	if resource.SubResource == downloadArchiveLink && method == http.MethodGet {
+		return true
+	}
+
+	if resource.SubResource == generateArchiveAction && method == http.MethodPost {
+		return true
+	}
```
Comment (m-ildefons) [inline pkg/api/upgradelog/handler.go]: Is there ever a use-case where we'd need the generate action but not the download action? If not, the generate action would be superfluous.

File: pkg/api/vm/handler.go [hunk 8]
```diff
@@ -95,11 +117,24 @@ func (h vmActionHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	rw.WriteHeader(http.StatusNoContent)
 }
 
-func (h *vmActionHandler) doAction(rw http.ResponseWriter, r *http.Request) error {
-	vars := util.EncodeVars(mux.Vars(r))
-	action := vars["action"]
-	namespace := vars["namespace"]
-	name := vars["name"]
+func (h *vmActionHandler) IsMatchedResource(resource subresource.Resource, method string) bool {
+	if resource.Name != subresource.VirtualMachines.Resource || method != http.MethodPost {
```
Comment (m-ildefons) [inline pkg/api/vm/handler.go]: Again, shouldn't we have HTTP method `PUT` for idempotent actions and HTTP method `GET` for listing-actions like `findMigratableNodes`?

**Review by m-ildefons — COMMENTED** [review]
Looks generally quite good. I still found a few places where I'm not sure about the usage of `POST` vs `PUT` and I think we should add a logging hook to make it easier to recognize actions when dealing with support cases.
I found the upgradelog API particularly confusing since I'm not sure why generating and downloading are two separate actions.

File: pkg/api/image/formatter.go (hunk 5 above)
Comment (Yu-Jack) [reply pkg/api/image/formatter.go]: What I'm doing is to migrate whole same functionality to sub resource handler, and original method is `POST`.  I didn't make too much changes to original behavior just in case, including HTTP method.

https://github.com/harvester/harvester/blob/bf8bdcd7423508f767bf389e43ee8e812639e01a/pkg/api/image/formatter.go#L71-L75

And another reason why it's `POST` because we're using rancher/steve to build our API server. In original documentation, `action` query string seems to only allow `POST` action ([reference](https://github.com/rancher/steve?tab=readme-ov-file#action)).

However, I was thinking and testing about this during these few days. I also think `PUT` would be better for those cases. So, I will use `PUT` method to replace all `POST` method if behavior is idempotent. I think this change would make our API look like more k8s standardized API. 

In short, I'll still check all behaviors are idempotent or not, and make it to `PUT` or `GET` method instead of all `POST` method.

**NOTE: For those people are using our action APIs, they might be confused about changing method. But, I think it's a reasonable change.**

File: pkg/api/node/formatter.go (hunk 6 above)
Comment (Yu-Jack) [reply pkg/api/node/formatter.go]: Same reason in https://github.com/harvester/harvester/pull/5704#discussion_r1622033858.

Sure, list-actions should be `GET`.

File: pkg/api/vm/handler.go (hunk 8 above)
Comment (Yu-Jack) [reply pkg/api/vm/handler.go]: Same reason in https://github.com/harvester/harvester/pull/5704#discussion_r1622033858.

File: pkg/api/image/formatter.go (hunk 5 above)
Comment (Yu-Jack) [reply pkg/api/image/formatter.go]: Actually, subresource concept is the verb to original resource. It's that we use subresources (verbs) to manipulate existing resource instead of creating, so it's basically a `PUT` or `GET` method. There shouldn't be `POST` method here.
